File: colle.cpp
#include <string>
#include <iostream>
#include <fstream>

using namespace std;

class colle
{
private:
 /* data */
 string fichier_acr;
 string merreur;

public:
 colle(/* args */ const string le_fichier);
 bool colle();
 string erreur_message()
 {
  return merreur;
 }
}

;

colle::colle(/* args */ const string le_fichier)
{
 fichier_acr = le_fichier;
}
bool colle::colle()
{
 string mon_fichier;
 string un_fichier;

 // ifstream in;
 ifstream inacr;
 ofstream out;
 int dim_total = 0;
 char tampon[512];
 int octets_lus = 0;
 int num_morceau = 0;
 int cs = 0;
 int cs1 = 0;
 int cs2 = 0;

 inacr.open(fichier_acr.c_str(), ios::in);
 if (inacr.fail() != 0)
 {
  merreur = "Erreur de lecture du fichier .acr";
  return;
 }
 inacr >> mon_fichier; // le fichier final
 if (mon_fichier.length() > 50)
 {
  merreur = "Fichier trop long (>50) ou erreur de lecture du fichier .acr";
  return;
 }
 out.open(mon_fichier.c_str(), ios::out | ios::binary);
 if (out.fail() != 0)
 {
  merreur = "Le fichier " + string(mon_fichier) + " ne peut être écrit";
  return false;
 }
 for (;;)
 {
  inacr >> un_fichier;
  if (un_fichier.length() < 1)
  {
   merreur = "Fichier .acr incorrect";
   return false;
  }
  if (un_fichier.compare("EnD") == 0)
  {
   out.close();
   inacr >> dim_total;

   if (dim_total == octets_trans)
    return true;
   else
   {
    merreur = "Nombre invalide de caractères transférés";
    return false;
   }
  }
  inacr >> cs;
  num_morceau++;
  cs1 = num_morceau;
  cs2 = 0;
  cout << "TEST: [" << un_fichier.c_str() << "]" << endl;

  ifstream in;
  in.open(un_fichier.c_str(), ios::in | ios::binary);
  if (in.fail() != 0)
  {
   merreur = "fichier " + string(un_fichier) + " ne peut être ouvert";
   return false;
  }
  for (;;)
  {
   in.read(tampon, 512);
   octets_lus = in.gcount();
   if (octets_lus <= 0)
    break;
   for (int j = 0; j < octets_lus; j++)
   {
    cs1 = cs1 ^ (int)tampon[j++] if (j != octets_lus)
    {
     cs2 = cs2 ^ (int)tampon[j];
    }
   }
   out.write(tampon, octets_lus);
   octets_trans += octets_lus;
  }
  in.close();
  if (cs != ((256 * cs1) + cs2))
  {
   merrreur = "Acr invalide pour le fichier " + string(un_fichier);
   return false;
  }
 }
}
int main(int argc, char const *argv[])
{
 string le_fichier;

 if (argc != 2)
 {
  cerr << "nombre de paramètres invalides" << endl;
  cerr << "Colle fichier_acr" << endl;

  le_fichier_arc = "prgm.acr";
  // return -1
 }
 else
 {
  le_fichier_arc = argv[1];
 }
 if (le_fichier_arc.rfind(".acr") != (le_fichier_arc.length() - 4))
 {
  cerr << "Ce n'est pas un fichier .acr" << endl;
  return -1;
 }
 colle ma_colle(le_fichier_arc);
 if (ma_colle.colle())
 {
  cout << "Colle terminée correctement  " << endl;
 }
 else
 {
  cerr << "Colle erreur: " << ma_colle.erreur_message() << endl;
 }
 return 0;
}

File: Colle.java
import java.io.*;

public class Colle {
 private String fichier_acr;
 private String merreur;

 public Colle(String le_fichier) {
  fichier_acr = le_fichier;
 }

 public boolean Colle() {
  String mon_fichier;
  String un_fichier;
  int dim_total = 0;
  byte[] tampon = new byte[512];
  int octets_lus = 0;
  int octets_trans = 0;
  int num_morceau = 0;
  int cs = 0;
  int cs1 = 0;
  int cs2 = 0;

  try {
   BufferedReader inacr = new BufferedReader(new FileReader(fichier_acr));
   try {
    mon_fichier = inacr.readLine(); // le fichier final
   } catch (IOException ioe) {
    merreur = " Erreur de lecture du fichier .acr";
    return false;
   }
   if (mon_fichier.length() > 50) {
    merreur = "Fichier trop long (>50 ) ou erreur de lecture du fichier .acr";
    return false;
   }
   try {
    FileOutputStream out = new FileOutputStream(new File(mon_fichier));
    try {
     for (;;) {
      un_fichier = inacr.readLine();
      if (un_fichier == null) {
       merreur = "Fichier .acr incorrect";
       return false;
      }
      if (un_fichier.equals("EnD")) {
       out.close();
       dim_total = Integer.parseInt(inacr.readLine());

       if (dim_total == octets_trans)
        return true;
       else {
        merreur = "Nombre invalide de caractères transférés";
        return false;
       }
      }
      cs = Integer.parseInt(inacr.readLine());
      num_morceau++;

      cs1 = num_morceau;
      cs2 = 0;

      FileInputStream in = new FileInputStream(un_fichier);
      try {
       for (;;) {
        octets_lus = in.read(tampon);
        if (octets_lus <= 0)
         break;
        for (int j = 0; j < octets_lus; j++) {
         cs1 = cs1 ^ (int) tampon[j++];
         if (j != octets_lus) {
          cs2 = cs2 ^ (int) tampon[j];
         }
        }
        out.write(tampon, 0, octets_lus);
        octets_trans += octets_lus;
       }
       in.close();
      } catch (IOException ioe) {
       merreur = "Erreur de lecture pour " + un_fichier;
       return false;
      }
      if (cs != ((256 * cs1) + cs2)) {
       merreur = "Acr invalide pour le fichier " + un_fichier;
       return false;
      }
     }
    } catch (IOException ioe) {
     merreur = "Fichier .acr invalide";
     return false;
    }
   } catch (FileNotFoundException e) {
    merreur = "Le fichier " + mon_fichier + " ne peut être écrit";
    return false;
   }

  } catch (Exception e) {
   merreur = "Le fichier " + fichier_acr + " n'existe pas!";
   return false;
  }
 }

 public String erreur_message() {
  return merreur;
 }

 public static void main(String[] args) {
  int args_len = args.length;

  if (args_len != 1) {
   System.err.println("Nombre de paramètres invalides");
   System.err.println("java Colle fichier_acr");
   return;
  }
  String le_fichier_acr = args[0];
  if (le_fichier_acr.lastIndexOf(".acr") != (le_fichier_acr.length() - 4)) {
   System.err.println("Ce n'est pas un fichier .acr");
   return;
  }
  Colle ma_colle = new Colle(le_fichier_acr);
  if (ma_colle.Colle()) {
   System.out.println("Colle erreur: " + ma_colle.erreur_message());
  } else {
   System.err.println("Colle erreur: " + ma_colle.erreur_message());
  }
 }
}
